pkg/distributor/distributor.go

```diff
@@ -440,40 +422,35 @@ func (d *Distributor) shardStream(stream logproto.Stream, streamSize int, userID
 		level.Info(logger).Log("msg", "sharding request", "shard_count", shardCount)
 	}
 
-	if shardCount > len(stream.Entries) {
-		shardsOrder := d.randomizeShardsOrder(shardCount)
-		return d.divideEntriesBetweenShards(logger, userID, len(stream.Entries), shardStreamsCfg, stream, shardsOrder)
-	}
-
-	return d.divideEntriesBetweenShards(logger, userID, shardCount, shardStreamsCfg, stream, nil)
+	return d.divideEntriesBetweenShards(userID, shardCount, shardStreamsCfg, stream)
 }
 
-func (d *Distributor) randomizeShardsOrder(shardCount int) []int {
-	shardsOrder := make([]int, shardCount)
-	for i := 0; i < shardCount; i++ {
-		shardsOrder = append(shardsOrder, i)
+func (d *Distributor) divideEntriesBetweenShards(userID string, totalShards int, shardStreamsCfg *shardstreams.Config, stream logproto.Stream) ([]uint32, []streamTracker) {
+	derivedKeys, derivedStreams := d.createShards(stream, totalShards, userID, shardStreamsCfg)
+
+	for i := 0; i < len(stream.Entries); i++ {
+		streamIndex := i % len(derivedStreams)
+		entries := append(derivedStreams[streamIndex].stream.Entries, stream.Entries[i])
+		derivedStreams[streamIndex].stream.Entries = entries
 	}
-	rand.Seed(time.Now().UnixNano())
-	rand.Shuffle(len(shardsOrder), func(i, j int) { shardsOrder[i], shardsOrder[j] = shardsOrder[j], shardsOrder[i] })
-	return shardsOrder
+
+	return derivedKeys, derivedStreams
 }
 
-func (d *Distributor) divideEntriesBetweenShards(logger log.Logger, userID string, shardCount int, shardStreamsCfg *shardstreams.Config, stream logproto.Stream, alternateShardsOrder []int) ([]uint32, []streamTracker) {
-	derivedKeys := make([]uint32, 0, shardCount)
-	derivedStreams := make([]streamTracker, 0, shardCount)
-	streamLabels := labelTemplate(stream.Labels)
-	streamPattern := streamLabels.String()
+func (d *Distributor) createShards(stream logproto.Stream, totalShards int, userID string, shardStreamsCfg *shardstreams.Config) ([]uint32, []streamTracker) {
+	var (
+		streamLabels   = labelTemplate(stream.Labels)
```

Comment:
I think `labelTemplate` needs to sort the labels after adding the shard label since we rely on labels to be sorted coming from the distributor.

Reply:
Already covered in https://github.com/grafana/loki/pull/7556